<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <!-- 1 -->
    <!-- <script>
        error
        console.log(1);
    </script>
    <script>
        // 会执行
        console.log(2);
    </script> -->

    <!-- 2 -->
    <!-- <script>
        try {
            console.log(error);
        } catch (e) {
            console.log(e);
        }
        // 异步的错误捕获不到
        try {
            setTimeout(() => {
                console.log(error);
            })
        } catch (e) {
            console.log('erroe:', e);
        }
    </script> -->

    <!-- 3 -->
    <!-- <script>
        try {
            setTimeout(() => {
                console.log(error);
            })
        } catch (e) {
            console.log('erroe:', e);
        }
        // 捕捉不到html中的错误
        window.onerror = function(msg, url, row, col, error) {
            console.log(msg);
            // 禁止页面报红
            return true
        }
    </script> -->

    <!-- 4 -->
    <!-- <img src="./xx.png" alt="">
    <script>
        window.addEventListener("error", (msg, url, row, col, error) => {
            console.log(msg);
            return true
        }, true)
    </script> -->
    <!-- 以上都捕获不到Promise的错误 -->

    <!-- 5 -->
    <script>
        // 捕获异步错误
        window.addEventListener('unhandledreject', function(e) {
            e.preventDefault()
            console.log(e.reason);
            return true
        })
        // 不catch就会报红
        new Promise((resolve, reject) => {
            reject("erroe1")
        })
        Promise.resolve().then(() => {
            throw "error3"
        })
        Promise.reject("error2")

        // 工具：fundebug
        // 接入
        // npm 安装

        // 用户录屏：sessionstack

        // 容错：zanePerfor （github）

        // lgwebdream
    </script>
</body>
</html>